src/builder/create_attachment.rs

```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+/// # async fn _foo(ctx: Context, mut msg: Message) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(EditAttachments::new())).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Adding a new attachment without deleting existing attachments
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).add(my_attachment)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete all but the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new().keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete only the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).dont_keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// # Notes
+///
+/// Internally, this type is used not just for message editing endpoints, but also for message
+/// creation endpoints.
+#[derive(Default, Debug, Clone, serde::Serialize)]
+#[serde(transparent)]
+#[must_use]
+pub struct EditAttachments {
+    new_and_existing_attachments: Vec<NewOrExisting>,
+}
+
+impl EditAttachments {
+    /// An empty attachments builder.
+    ///
+    /// Existing attachments are not kept by default, either. See [`Self::new_keep_all()`] or
+    /// [`Self::keep()`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new attachments builder that keeps all existing attachments.
+    ///
+    /// Shorthand for [`Self::new()`] and calling [`Self::keep()`] for every
+    /// [`AttachmentId`] in [`Message::attachments`].
+    ///
+    /// If you only want to keep a subset of attachments from the message, either implement this
+    /// method manually, or use [`Self::dont_keep()`].
+    ///
+    /// **Note: this EditAttachments must be run on the same message as is supplied here, or else
+    /// Discord will throw an error!**
+    pub fn new_keep_all(msg: &Message) -> Self {
+        Self {
+            new_and_existing_attachments: msg
+                .attachments
+                .iter()
+                .map(|a| {
+                    NewOrExisting::Existing(ExistingAttachment {
+                        id: a.id,
+                    })
+                })
+                .collect(),
+        }
+    }
+
+    /// This method adds an existing attachment to the list of attachments that are kept after
+    /// editing.
+    pub fn keep(mut self, id: AttachmentId) -> Self {
```

Comment:
What do you think of the name `retain`?

Reply:
Not a fan tbh. In Rust I associate it with `collection.retain(<closure>)`, which this is not. Also, I'm always a fan if method names match natural language, and I'd naturally say "keep an attachment" rather than "retain an attachment" (for example in the docs of that method)